robot/camera.py

```diff
@@ -1,9 +1,30 @@
-import socket
-import time
 from pathlib import Path
-
 from robot.board import Board
 from robot.markers import Marker
+import time
+from typing import List
+import socket
+
+
+class ResultList(List[Marker]):
+    """
+    A ``list`` class with nicer error messages.
+
+    In particular, this class provides a slightly better error description when
+    accessing indexes and the list is empty.
+
+    This is to mitigate a common beginners issue where a list is indexed
+    without checking that the list has any items.
+    """
+
+    def __getitem__(self, index: int) -> Marker:
+        try:
+            return super().__getitem__(index)
+        except IndexError as e:
+            if not self:
+                raise IndexError("Trying to index an empty list") from None
+            else:
+                raise e
```

Comment:
minor: this should probably be `raise` rather than `raise e`

Reply:
The functionality is the same, but I prefer to be explicit if we're extracting the exception as a variable anyway

Reply:
In Python 2.7 the functionality between the two was very different; I've not yet had enough of a play with Python 3 to know what the differences are now.

Even if thy are the same, I think I'll still be in favour of plain `raise` since that's the idiomatic way of saying "I'm re-raising the original exception, as if I'd not caught it" (which is what this code is trying to do).

Reply:
Hrm, so it turns out that in Python 3 we don't lose the underlying trace (like we did in Python 2), though there is a difference in the outcome from re-raising vs raising the original exception explicitly. Consider the following example:
``` python
import traceback

def func():
    raise Exception("from func")

def reraise():
    try:
        func()
    except Exception as e:
        raise

def raise_e():
    try:
        func()
    except Exception as e:
        raise e

for fn in (reraise, raise_e):
    print("\nTesting with {}".format(fn.__name__ ))
    try:
        fn()
    except:
        traceback.print_exc()
```
The output of this is:
```
> py foo.py

Testing with reraise
Traceback (most recent call last):
  File "foo.py", line 21, in <module>
    fn()
  File "foo.py", line 8, in reraise
    func()
  File "foo.py", line 4, in func
    raise Exception("from func")
Exception: from func

Testing with raise_e
Traceback (most recent call last):
  File "foo.py", line 21, in <module>
    fn()
  File "foo.py", line 16, in raise_e
    raise e
  File "foo.py", line 14, in raise_e
    func()
  File "foo.py", line 4, in func
    raise Exception("from func")
Exception: from func

> py --version
Python 3.6.4
```

Reply:
Looks like we don't use the underlying exception (something I didn't know python2 did, which just sounds annoying). But seeing the stack trace there, it'd be nice not to show the additional `raise` statement. Will revert back to a plain `raise`.

Reply:
Resolved in #59 by 960c531.